sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -126,6 +126,149 @@ case class MapKeys(child: Expression)
+
+  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.length)(ArrayType)
+
+  override def dataType: DataType = ArrayType(mountSchema)
+
+  override def prettyName: String = "zip"
+
+  override def nullable: Boolean = children.forall(_.nullable)
+
+  lazy val numberOfArrays: Int = children.length
+
+  private lazy val arrayTypes = children.map(_.dataType.asInstanceOf[ArrayType])
+
+  private lazy val arrayElementTypes = arrayTypes.map(_.elementType)
+
+  def mountSchema: StructType = {
+    val fields = arrayTypes.zipWithIndex.foldRight(List[StructField]()) { case ((arr, idx), list) =>
+      StructField(s"_$idx", arr.elementType, children(idx).nullable || arr.containsNull) :: list
+    }
+    StructType(fields)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val genericArrayData = classOf[GenericArrayData].getName
+    val genericInternalRow = classOf[GenericInternalRow].getName
+
+    val evals = children.map(_.genCode(ctx))
+
+    val arrVals = ctx.freshName("arrVals")
+    val arrCardinality = ctx.freshName("arrCardinality")
+    val biggestCardinality = ctx.freshName("biggestCardinality")
+    val storedArrTypes = ctx.freshName("storedArrTypes")
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+        |${eval.code}
+        |if (!${eval.isNull}) {
+        |  $arrVals[$index] = ${eval.value};
+        |  $arrCardinality[$index] = ${eval.value}.numElements();
+        |} else {
+        |  $arrVals[$index] = null;
+        |  $arrCardinality[$index] = 0;
+        |}
+        |$storedArrTypes[$index] = "${arrayElementTypes(index)}";
+        |$biggestCardinality[0] = Math.max($biggestCardinality[0], $arrCardinality[$index]);
+      """.stripMargin
+    }
+
+    val inputsSplitted = ctx.splitExpressions(
+      expressions = inputs,
+      funcName = "getInputAndCardinality",
+      arguments =
+        ("ArrayData[]", arrVals) ::
+        ("int[]", arrCardinality) ::
+        ("String[]", storedArrTypes) ::
+        ("int[]", biggestCardinality) :: Nil)
+
+    val myobject = ctx.freshName("myobject")
+    val j = ctx.freshName("j")
+    val i = ctx.freshName("i")
+    val args = ctx.freshName("args")
+
+    val fillValue = arrayElementTypes.distinct.map { case (elementType) =>
+      val getArrValsItem = CodeGenerator.getValue(s"$arrVals[$j]", elementType, i)
+      s"""
+      |      if ($storedArrTypes[$j] == "${elementType}") {
+      |        $myobject[$j] = $getArrValsItem;
+      |      }
+      """.stripMargin
+    }
+
+    val fillValueSplitted = ctx.splitExpressions(
+      expressions = fillValue,
+      funcName = "fillValue",
+      arguments =
+        ("int", j) ::
+        ("Array[] Object", myobject) :: Nil)
+
+    ev.copy(s"""
+      |ArrayData[] $arrVals = new ArrayData[$numberOfArrays];
+      |int[] $arrCardinality = new int[$numberOfArrays];
+      |int[] $biggestCardinality = new int[1];
```

Comment:
as we said this should be a int

Reply:
I used `int[]` because I thought that it was impossible to change $biggestCardinality due to java not allowing passing int by reference. But I found a workaround (inspired by `ConcatWs`).

Reply:
hint: the right approach here is to return it as result of the splitted functions